File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int solverRec(TreeNode* root, int &dia){
        if(root==NULL){
            return 0;
        }
        int lh = solverRec(root->left,dia);
        int rh = solverRec(root->right,dia);
        dia = max(dia, lh+rh);
        return max(lh,rh)+1; 
    
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int dia =0;
        solverRec(root,dia);
        return dia;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    dia =0
    def solverRec(self,root):
        if root==None:
            return 0
        
        lh = self.solverRec(root.left)
        rh = self.solverRec(root.right)

        self.dia = max(self.dia, lh+rh)

        return max(lh,rh)+1
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        self.solverRec(root)
        return self.dia
        
        